core/src/main/java/org/apache/calcite/rel/metadata/RelMetadataQueryBase.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.metadata;
+
+import org.apache.calcite.rel.RelNode;
+
+import java.lang.reflect.Proxy;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+/**
+ * Base class for the RelMetadataQuery that uses the metadata handler class
+ * generated by the Janino.
+ *
+ * <p>To add a new implementation to this interface, follow
+ * these steps:
+ *
+ * <ol>
+ * <li>Extends {@link RelMetadataQuery} (name it MyRelMetadataQuery for example)
+ * to reuse the Calcite builtin metadata query interfaces. In this class, define all the
+ * extended Handlers for your metadata and implement the metadata query interfaces.
+ * <li>Write your customized provider class <code>RelMdXyz</code>. Follow
+ * the pattern from an existing class such as {@link RelMdColumnOrigins},
+ * overloading on all of the logical relational expressions to which the query
+ * applies.
+ * <li>Add a {@code SOURCE} static member to each of your provider class, similar to
+ * {@link RelMdColumnOrigins#SOURCE}.
+ * <li>Extends {@link DefaultRelMetadataProvider} (name it MyRelMetadataProvider for example)
+ * and supplement the "SOURCE"s into the builtin list
+ * (This is not required, use {@link ChainedRelMetadataProvider} to chain your customized
+ * "SOURCE"s with default ones also works).
+ * <li>Set {@code MyRelMetadataProvider} into the cluster instance.
+ * <li>Use
+ * {@link org.apache.calcite.plan.RelOptCluster#setMetadataQuery(Supplier)}
+ * to set the metadata query {@link Supplier} into the cluster instance. This {@link Supplier}
+ * should return a <strong>fresh new</strong> instance.
+ * <li>Use the cluster instance to create
+ * {@link org.apache.calcite.sql2rel.SqlToRelConverter}.</li>
+ * <li>Query your metadata within {@link org.apache.calcite.plan.RelOptRuleCall} with the
+ * interfaces you defined in {@code MyRelMetadataQuery}.
+ * </ol>
+ */
+public class RelMetadataQueryBase {
```

Comment:
I mean, remove this class. I know you are following Julian's suggestion, which I don't agree.

Reply:
Sorry, i still think Julian's suggestion make the definition more clear.

Reply:
You are adding a base class no one ever is going to extend, except RelMetadataQuery.

Reply:
They actually can extends RelMetadataQueryBase directly if they do not want the handlers defined in `RelMetadataQuery`.

Reply:
How can? Given that you already defined the API as:
```
public <M extends RelMetadataQuery> M getMetadataQuery();
public void setMetadataQuery(Supplier<RelMetadataQuery> mqSupplier);
```

Reply:
Well, exactly, i think the defined API should use `RelMetadataQuery` instead. The `RelMetadataQueryBase` behaviors like the Scala trait, but Java does not have trait, do you have any good suggestions ?

Reply:
Remove it, it is an over-design.

Reply:
@julianhyde What do you think about this base class ?

Reply:
I don't think it's over-design. The code in the base class is doing different work than the code in the derived class.

Maybe it doesn't have to be a base class. Maybe it should be an embedded object instead. How do you feel about that?

Reply:
Yes, instead of inheritance, we can use composition, let me have a try

Reply:
I am fine with it. We can keep it as a base class.

Reply:
Okey, after a try, i also think the inheritance is much more intuitive than composition, let's keep it as it is now.